### PR TITLE
Shipping Labels: state machine, address validation in shipping labels form, suggested address logics

### PR DESCRIPTION
**Pull request by pmusolino**
Part of #2973

## Description
In this PR, many things have been handled under the hood, so the changes visible to the user are few.
- [x] the validation address endpoint will be called also in the main Shipping Label Form, for validating the origin/destination address when the user presses **Continue** button.
- [x] If the address is validated, the next step in the shipping label form will be enabled (implemented the state machine).
- [x] If there is a suggested address, the new screen that will allow the users to choose the suggested address will be shown (WIP, will be completed in the next PR).
- [x] If the user edits an existing address, after the validation the next step in the shipping label form will be enabled.

## Changes
- Implemented a new `UIButton` subclass called `ButtonActivityIndicator` that is able to display or hide an activity indicator in the middle.
- The new button `ButtonActivityIndicator` is now implemented inside the `ShippingLabelFormStepTableViewCell`, for showing an activity indicator on top of the `Continue` button when we are validating an address.
- `ShippingLabelSuggestedAddressTopBannerFactory`: the class that will return the top banner view inside the Shipping Label Suggested Address screen.
- Started to implement the `ShippingLabelSuggestedAddressViewController`. This is still a WIP and will be completed in the next PR.
- `ShippingLabelAddressFormViewController`: handled all the navigation things and removed the `topBanner` row which was unused. Implemented also the validation address for the origin and destination address, with the related navigation flow.
- `ShippingLabelFormViewModel`: declared a new struct called `State` which will handle all the necessary state to produce the ViewModel's outputs, including the sections. Implemented also the state machine flow to handle all the changes when row data is validated. Implemented the endpoint for validating an address directly in the main shipping label form.
- Added new tests.

## Testing
1. Launch the app in debug mode (because of the feature flag).
2. Open an order detail.
3. Press the button "Create Shipping Label".
4. Press the button "Continue" in the Create Shipping Label form. -> The activity indicator on top of the Continue button should appear.
5. If the address is valid, you should see the destination address with the **Continue** button enabled (in the origin cell the Continue button should disappear). If the address is not valid, you will navigate to the editing screen. If the address is valid but has a suggested address, you will navigate to the new screen with the suggested address (WIP).

## Screenshots
| Continue button pressed in Shipping Label Form           |  Shipping Label Suggested Address screen (WIP) |
:-------------------------:|:-------------------------:
Simulator Screen Shot - iPhone 12 Pro - 2021-03-03 at 15 53 45 | Simulator Screen Shot - iPhone 12 Pro - 2021-03-03 at 15 53 48

Update release notes:
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/Create Shipping Label Form/ShippingLabelFormViewController.swift [hunk 1]
```diff
@@ -110,16 +118,37 @@ private extension ShippingLabelFormViewController {
...
+                    let shippingAddressVC = ShippingLabelAddressFormViewController(siteID: self.viewModel.siteID,
+                                                                                   type: .origin,
+                                                                                   address: shippingLabelAddress,
+                                                                                   completion: { [weak self] (newShippingLabelAddress) in
+                                                                                    guard let self = self else { return }
+                            self.viewModel.handleOriginAddressValueChanges(address: newShippingLabelAddress, validated: true)
+
+                                                                                   })
+                    self.navigationController?.pushViewController(shippingAddressVC, animated: true)
+                }
```
Comment (ealeksandrov) [inline WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/Create Shipping Label Form/ShippingLabelFormViewController.swift:132]: nit: 2 levels of callback closures are hard to read. Maybe we can move code for VC inits and callbacks into separate methods? Like:
```swift
case .suggestedAddress:
    self.showSuggestedAddressVC(for: shippingLabelAddress, suggestedAddress: response?.address)
```

File: WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/Create Shipping Label Form/ShippingLabelFormViewController.swift [hunk 2]
```diff
@@ -129,16 +158,37 @@ private extension ShippingLabelFormViewController {
...
+                    let shippingAddressVC = ShippingLabelAddressFormViewController(siteID: self.viewModel.siteID,
+                                                                                   type: .destination,
+                                                                                   address: shippingLabelAddress,
+                                                                                   completion: { [weak self] (newShippingLabelAddress) in
+                                                                                    guard let self = self else { return }
+                            self.viewModel.handleDestinationAddressValueChanges(address: newShippingLabelAddress, validated: true)
+
+                                                                                   })
+                    self.navigationController?.pushViewController(shippingAddressVC, animated: true)
+                }
```
Comment (ealeksandrov) [inline WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/Create Shipping Label Form/ShippingLabelFormViewController.swift:159]: nit: decouple VC navigation like in `configureShipFrom` suggestion above, maybe reuse same methods if that will be clean.

File: WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/Create Shipping Label Form/ShippingLabelFormViewModel.swift [hunk 3]
```diff
@@ -97,4 +164,62 @@ private extension ShippingLabelFormViewModel {
         try? resultsController.performFetch()
         return resultsController.fetchedObjects.first
     }
+
+    func isValidatingAddress(_ validating: Bool, type: ShipType) {
```
Comment (ealeksandrov) [inline WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/Create Shipping Label Form/ShippingLabelFormViewModel.swift]: nit: make more obvious that it's a function to change an internal state, like:

```suggestion
    func updateValidatingAddressState(_ validating: Bool, type: ShipType) {
```

File: WooCommerce/Classes/ViewRelated/ReusableViews/ButtonActivityIndicator.swift [hunk 4]
```diff
@@ -0,0 +1,56 @@
+import UIKit
+
+/// Displays a button with an activity indicator in the center
+///
+final class ButtonActivityIndicator: UIButton {
+
+    private let indicator: UIActivityIndicatorView = UIActivityIndicatorView(style: UIActivityIndicatorView.Style.medium)
```
Comment (ealeksandrov) [inline WooCommerce/Classes/ViewRelated/ReusableViews/ButtonActivityIndicator.swift]: nit:
```suggestion
    private let indicator: UIActivityIndicatorView = UIActivityIndicatorView(style: .medium)
```

File: WooCommerce/Classes/ViewRelated/ReusableViews/ButtonActivityIndicator.swift [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    ///
+    func hideActivityIndicator() {
+        guard subviews.contains(indicator) == true else {
+            return
+        }
+        indicator.stopAnimating()
+        indicator.removeFromSuperview()
+        UIView.transition(with: self, duration: Constants.animationDuration, options: .curveEaseIn, animations: { [weak self] in
+            self?.titleLabel?.alpha = 1.0
+        }) { (finished) in
```
Comment (ealeksandrov) [inline WooCommerce/Classes/ViewRelated/ReusableViews/ButtonActivityIndicator.swift:47]: nit:
```suggestion
        }) { _ in
```

File: WooCommerce/WooCommerceTests/ViewRelated/Shipping Label/Create Shipping Label/ShippingLabelFormViewModelTests.swift [hunk 6]
```diff
@@ -71,9 +71,72 @@ final class ShippingLabelFormViewModelTests: XCTestCase {
...
+        // When
+        XCTAssertEqual(currentRows?[0].type, .shipFrom)
+        XCTAssertEqual(currentRows?[0].dataState, .pending)
+        XCTAssertEqual(currentRows?[0].displayMode, .editable)
+        XCTAssertEqual(currentRows?[1].type, .shipTo)
+        XCTAssertEqual(currentRows?[1].dataState, .pending)
+        XCTAssertEqual(currentRows?[1].displayMode, .disabled)
+        XCTAssertEqual(currentRows?[2].type, .packageDetails)
+        XCTAssertEqual(currentRows?[2].dataState, .pending)
+        XCTAssertEqual(currentRows?[2].displayMode, .disabled)
```
Comment (ealeksandrov) [inline WooCommerce/WooCommerceTests/ViewRelated/Shipping Label/Create Shipping Label/ShippingLabelFormViewModelTests.swift:93]: nit: maybe asserts for initial state should belong to `Given` section, so `When` is more focused on action being tested?

**Review by ealeksandrov — APPROVED** [review]
LGTM 👏 
Only some nits left for your consideration.

File: WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/Create Shipping Label Form/ShippingLabelFormViewController.swift (hunk 1 above)
Comment (pmusolino) [reply WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/Create Shipping Label Form/ShippingLabelFormViewController.swift:132]: Sounds good to me 👍 269345410d8a3d573635ad5ad6b7f7d87f04f581

File: WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/Create Shipping Label Form/ShippingLabelFormViewController.swift (hunk 2 above)
Comment (pmusolino) [reply WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/Create Shipping Label Form/ShippingLabelFormViewController.swift:159]: ✅  a79b10b3c29233f5525a598f65bcc3d84aab081c

File: WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/Create Shipping Label Form/ShippingLabelFormViewModel.swift (hunk 3 above)
Comment (pmusolino) [reply WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/Create Shipping Label Form/ShippingLabelFormViewModel.swift]: 👍 ec4fba661839938985e0d23a3ec25bd39cd3447e great suggestion

File: WooCommerce/WooCommerceTests/ViewRelated/Shipping Label/Create Shipping Label/ShippingLabelFormViewModelTests.swift (hunk 6 above)
Comment (pmusolino) [reply WooCommerce/WooCommerceTests/ViewRelated/Shipping Label/Create Shipping Label/ShippingLabelFormViewModelTests.swift:93]: yes, it sounds right 👍 Updated here ab83f198ac68d0eae0799b09a89e8cbd8b047203

**Comment by pmusolino** [comment]
Thank you 💯 for the review @ealeksandrov 😉 I'm going to merge it after the CI is ✅

**Pull request merged**